docs: guides section (#717)
software-mansion/react-native-enriched: docs/docs/guides/user-and-channel-mentions.mdx

---

> sidebar_position: 2

Reviewer: ```suggestion
sidebar_position: 1
```

# User and channel mentions

This guide wires the mention events into a complete picker - the flow behind a
chat composer where typing `@` suggests people and `#` suggests channels. If you
haven't met the mention API yet, read
[Mentions](/rich-text-formatting/mentions) first; here we move quickly and
assume the events and `setMention` are familiar.

## The data behind a mention

Mentions are particularly useful if they point at something. Keep two lists around - one
for users, one for channels - each item carrying an `id` you can attach
to the finished mention:

```tsx
const USERS = [
  { id: 'u1', name: 'John Doe' },
  { id: 'u2', name: 'Jane Smith' },
  { id: 'u3', name: 'Alice Johnson' },
  { id: 'u4', name: 'Bob Brown' },
];

const CHANNELS = [
  { id: 'c1', name: 'general' },
  { id: 'c2', name: 'engineering' },
  { id: 'c3', name: 'random' },
  { id: 'c4', name: 'announcements' },
];
```

## Wiring the picker

Now let's register both indicators and the events callbacks:

```tsx
<EnrichedTextInput
  mentionIndicators={['@', '#']}
  onStartMention={openPicker} // fired when '@' or '#' is typed

>   onChangeMention={updateQuery} // fired on every keystroke after it

Reviewer 2: ```suggestion
  onChangeMention={updateQuery} // fired on every keystroke after an indicator
```

  onEndMention={closePicker} // fired when the mention is left
  // ...
/>
```

`onStartMention` hands you the indicator, so you know whether to show people or
channels. `onChangeMention` hands you the `text` typed so far - filter your list
with it. `onEndMention` fires when the cursor leaves the mention, so you dismiss
the list.

When the user taps a suggestion, finish the mention with `setMention`. Pass the
same indicator that started it, the display text, and
the item's data as attributes:

```tsx
const pick = (item) => {
  ref.current?.setMention(indicator, `${indicator}${item.name}`, {
    id: item.id,
  });
};
```

Now let's give each indicator its own look through
`htmlStyle.mention`:

```tsx
const htmlStyle = {
  mention: {
    '@': { color: '#2b7a4b', backgroundColor: '#d8f3e3' },
    '#': { color: '#2b5f9e', backgroundColor: '#d8e6f9' },
  },
};
```

:::info

If you want to see the whole code used to build this example, you can find it below by switching the tab from `Preview` to `Code`.

:::

## Try it out

Type `@` to filter people or `#` to filter channels, keep typing to narrow the
list, then tap a suggestion.

> <InteractiveExample src={MentionOnlyEditorSrc} component={MentionOnlyEditor} />

Reviewer: There is some weird jumping on left menu when we are opening mentions picker:

https://github.com/user-attachments/assets/27e9eeb2-2fa2-434f-91d3-22f0edb287a9

Author: It's a docusaurus issue, not worth fixing, at least currently

> :::

Reviewer 3: We don't need this explanation, we already have those snippets everywhere in the docs, so user should be familiar with that. 
```suggestion
```
